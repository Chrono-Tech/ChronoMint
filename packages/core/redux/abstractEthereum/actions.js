/**
 * Copyright 2017â€“2019, LaborX PTY
 * Licensed under the AGPL Version 3 license.
 */

import {
  NONCE_UPDATE,
} from './constants'

export const nonceUpdate = (blockchain) => (address, nonce) => ({
  type: NONCE_UPDATE,
  address,
  blockchain,
  nonce,
})
