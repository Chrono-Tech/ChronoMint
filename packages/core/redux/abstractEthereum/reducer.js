/**
 * Copyright 2017â€“2019, LaborX PTY
 * Licensed under the AGPL Version 3 license.
 */

import {
  NONCE_UPDATE,
} from './constants'

import {
  BLOCKCHAIN_ETHEREUM,
  BLOCKCHAIN_LABOR_HOUR,
} from '../../dao/constants'

const initialSubState = {
  nonces: {},
}

const initialState = () => ({
  [BLOCKCHAIN_ETHEREUM]: {
    ...initialSubState,
  },
  [BLOCKCHAIN_LABOR_HOUR]: {
    ...initialSubState,
  },
})

const mutations = {
  [NONCE_UPDATE]: (state, { address, blockchain, nonce }) => {
    const blockchainScope = state[blockchain]
    const nonces = blockchainScope.nonces

    return ({
      ...state,
      [blockchain]: {
        ...blockchainScope,
        nonces: {
          ...nonces,
          [address]: nonce,
        },
      },
    })
  },
}

export default (state = initialState(), { type, ...payload }) => {
  return (type in mutations)
    ? mutations[type](state, payload)
    : state
}
