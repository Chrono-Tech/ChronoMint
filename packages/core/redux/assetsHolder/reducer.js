/**
 * Copyright 2017â€“2018, LaborX PTY
 * Licensed under the AGPL Version 3 license.
 */

import AssetHolderModel from '../../models/assetHolder/AssetHolderModel'
import * as a from './constants'

const initialState = new AssetHolderModel()

export default (state = initialState, action) => {
  switch (action.type) {
    case a.ASSET_HOLDER_INIT:
      return state.isInited(true)
    case a.ASSET_HOLDER_ADDRESS:
      return state
        .account(action.account)
        .wallet(action.wallet)
    case a.ASSET_HOLDER_ASSET_UPDATE:
      return state.assets(state.assets().update(action.asset))
    default:
      return state
  }
}
