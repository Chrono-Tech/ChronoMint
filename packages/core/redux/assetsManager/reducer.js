/**
 * Copyright 2017â€“2018, LaborX PTY
 * Licensed under the AGPL Version 3 license.
 */

import MainAssetsManagerModel from '../../models/assetsManager/MainAssetsManagerModel'
import {
  GET_ASSETS_MANAGER_COUNTS,
  GET_ASSETS_MANAGER_COUNTS_START,
  GET_PLATFORMS,
  GET_TRANSACTIONS_DONE,
  GET_TRANSACTIONS_START,
  SELECT_PLATFORM,
  SELECT_TOKEN,
  SET_ASSETS,
  SET_NEW_MANAGERS_LIST,
  GET_ASSET_DATA,
} from './constants'

export const initialState = new MainAssetsManagerModel()

export default (state = initialState, action) => {
  switch (action.type) {
    case GET_ASSETS_MANAGER_COUNTS_START:
      return state.isFetching(true).isFetched(false)
    case GET_ASSETS_MANAGER_COUNTS:
      return state
        .isFetching(false)
        .isFetched(true)
        .managersCount(action.payload.managers.length || 0)
        .assets(action.payload.assets)
        .platformsList(action.payload.platforms)
        .usersPlatforms(action.payload.usersPlatforms)
    case GET_ASSET_DATA:
      return state
        .assets({
          ...state.assets(),
          [ action.asset.token ]: action.asset,
        })
    case GET_PLATFORMS:
      return state
        .platformsList(action.payload.platforms)
        .usersPlatforms(action.payload.usersPlatforms)
    case SELECT_TOKEN:
      return state.selectedToken(action.payload.symbol)
    case SELECT_PLATFORM:
      return state.selectedPlatform(action.payload.platformAddress)
    case SET_ASSETS:
      return state.assets(action.payload.assets)
    case SET_NEW_MANAGERS_LIST:
      return state.managersCount(action.payload.managers.length)
    case GET_TRANSACTIONS_START:
      return state
        .transactionsList(
          state.transactionsList()
            .isFetched(false)
            .isFetching(true),
        )
    case GET_TRANSACTIONS_DONE:
      return state
        .transactionsList(
          state.transactionsList()
            .merge(action.payload.transactionsList)
            .isFetched(true)
            .isFetching(false),
        )
    default:
      return state
  }
}
