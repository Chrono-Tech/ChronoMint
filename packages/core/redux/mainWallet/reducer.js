/**
 * Copyright 2017â€“2018, LaborX PTY
 * Licensed under the AGPL Version 3 license.
 */

import { REHYDRATE } from 'redux-persist'
import MainWalletModel from '../../models/wallet/MainWalletModel'
import TransactionsCollection from '../../models/wallet/TransactionsCollection'
import * as a from './constants'

const initialState = new MainWalletModel()

export default (state = initialState, action) => {
  switch (action.type) {
    case REHYDRATE: {
      const incoming = action.payload.mainWallet
      if (incoming && incoming instanceof MainWalletModel) return state.names(incoming.names())
      return state
    }
    case a.WALLET_INIT:
      return state.isInited(action.isInited)
    case a.WALLET_BALANCE: // TODO @ipavlenko: Odd code, remove WALLET_BALANCE
      return state.balances(state.balances().update(
        state.balances().item(action.token.id()).updateBalance(action.isCredited, action.amount),
      ))
    case a.WALLET_ALLOWANCE:
      return state.allowances(state.allowances().update(action.allowance))
    case a.WALLET_ADDRESS:
      return state.addresses(state.addresses().update(action.address))
    case a.WALLET_TRANSACTIONS_FETCH:
      return state.updateTransactionsGroup({
        blockchain: action.blockchain,
        address: action.address,
        group: (state.transactions({
          blockchain: action.blockchain,
          address: action.address,
        }) || new TransactionsCollection()).isFetching(true),
      })
    case a.WALLET_TRANSACTION:
      return state.setTransaction(action.tx)
    case a.WALLET_TRANSACTION_UPDATED:
      return state.setTransaction(action.tx)
    case a.WALLET_TRANSACTIONS:
      return state.updateTransactionsGroup({
        blockchain: action.blockchain,
        address: action.address,
        group: action.group,
      })
    case a.WALLET_IS_TIME_REQUIRED:
      return state.isTIMERequired(action.value)
    case a.WALLET_TOKEN_BALANCE:
      return state.balances(state.balances().update(
        action.balance,
      ))
    case a.WALLET_SET_NAME:
      return state.names(state.names().set(`${action.blockchain}-${action.address}`, action.name))

    default:
      return state
  }
}
