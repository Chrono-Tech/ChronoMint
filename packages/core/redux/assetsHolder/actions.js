/**
 * Copyright 2017â€“2018, LaborX PTY
 * Licensed under the AGPL Version 3 license.
 */

import Amount from '../../models/Amount'
import AssetModel from '../../models/assetHolder/AssetModel'
import TokenModel from '../../models/tokens/TokenModel'
import AllowanceModel from '../../models/wallet/AllowanceModel'
import { DUCK_SESSION } from '../session/constants'
import { subscribeOnTokens } from '../tokens/actions'
import tokenService from '../../services/TokenService'
import { daoByType } from '../daos/selectors'
import TxExecModel from '../../models/TxExecModel'
import { getWallet } from '../wallets/selectors/models'
import { WALLETS_UPDATE_WALLET } from '../wallets/constants'
import WalletModel from '../../models/wallet/WalletModel'
import AllowanceCollection from '../../models/AllowanceCollection'

import {
  DUCK_ASSETS_HOLDER,
  ASSET_HOLDER_INIT,
  ASSET_HOLDER_ADDRESS,
  ASSET_HOLDER_ASSET_UPDATE,
} from './constants'

const handleToken = (token: TokenModel) => async (dispatch, getState) => {
  const assetHolder = getState().get(DUCK_ASSETS_HOLDER)
  const assets = assetHolder.assets()

  if (!token.isERC20() || !assets.list().has(token.address())) {
    return
  }

  const holderWallet = assetHolder.wallet()

  // set symbol for asset
  const asset = assets.item(token.address()).symbol(token.symbol())
  dispatch({ type: ASSET_HOLDER_ASSET_UPDATE, asset })

  // subscribe to token
  const tokenDAO = tokenService.getDAO(token.id())
  tokenDAO
    .on('approval', (results) => {
      if (results.from === holderWallet || results.spender === holderWallet) {
        dispatch(fetchAssetAllowance(token))
      }
    })
    .on('transfer', (data) => {
      if (!(data.from === holderWallet || data.to === holderWallet)) {
        return
      }
      dispatch(fetchAssetDeposit(token))
      dispatch(fetchAssetAllowance(token))
    })

  // need to be uncomment
  // const holderAccount = assetHolder.account()
  // await tokenDAO.watch(holderAccount)

  // fetch deposit and allowance
  dispatch(fetchAssetDeposit(token))
  dispatch(fetchAssetAllowance(token))
}

export const fetchAssetDeposit = (token: TokenModel) => async (dispatch, getState) => {
  const { account } = getState().get(DUCK_SESSION)
  const assetHolderDAO = daoByType('TimeHolder')(getState())
  const deposit = await assetHolderDAO.getDeposit(token.address(), account)
  const asset = getState().get(DUCK_ASSETS_HOLDER).assets().item(token.address()).deposit(new Amount(
    deposit,
    token.symbol(),
  ))
  dispatch({ type: ASSET_HOLDER_ASSET_UPDATE, asset })
}

export const fetchAssetAllowance = (token: TokenModel) => async (dispatch, getState) => {
  const assetHolder = getState().get(DUCK_ASSETS_HOLDER)
  const { account } = getState().get(DUCK_SESSION)

  const holderWallet = assetHolder.wallet()
  const tokenDAO = tokenService.getDAO(token.id())
  const assetHolderWalletAllowance = await tokenDAO.getAccountAllowance(account, holderWallet)

  const wallet = getWallet(`Ethereum-${account}`)(getState())
  const allowance = new AllowanceModel({
    amount: new Amount(assetHolderWalletAllowance, token.id()),
    spender: holderWallet,
    token: token.id(),
    isFetching: false,
    isFetched: true,
  })

  dispatch({
    type: WALLETS_UPDATE_WALLET,
    wallet: new WalletModel({
      ...wallet,
      allowances: new AllowanceCollection({
        list: {
          ...wallet.allowances.list,
          [allowance.id()]: allowance,
        },
      }),
    }),
  })
}

export const initAssetsHolder = () => async (dispatch, getState) => {
  if (getState().get(DUCK_ASSETS_HOLDER).isInited()) {
    return
  }
  dispatch({ type: ASSET_HOLDER_INIT, inInited: true })

  const assetHolderDAO = daoByType('TimeHolder')(getState())
  const [walletAddress] = await Promise.all([
    assetHolderDAO.getWalletAddress(),
  ])

  dispatch({ type: ASSET_HOLDER_ADDRESS, account: assetHolderDAO.address, wallet: walletAddress.toLowerCase() })

  // get assets list
  const [timeAddress] = await Promise.all([
    assetHolderDAO.getSharesContract(),
  ])
  const assets = [timeAddress]
  assets.forEach((address) => {
    dispatch({
      type: ASSET_HOLDER_ASSET_UPDATE,
      asset: new AssetModel({
        address: address.toLowerCase(),
      }),
    })
  })

  dispatch(subscribeOnTokens(handleToken))
}

export const depositAsset = (amount: Amount, token: TokenModel, feeMultiplier: Number = 1, advancedOptions = undefined) => async (dispatch, getState) => {
  try {
    const assetHolderDAO = daoByType('TimeHolder')(getState())
    const { account } = getState().get(DUCK_SESSION)
    advancedOptions['account'] = account
    await assetHolderDAO.deposit(token, amount, feeMultiplier, advancedOptions)
  } catch (e) {
    // eslint-disable-next-line
    console.error('deposit error', e)
  }
}

export const withdrawAsset = (amount: Amount, token: TokenModel, feeMultiplier: Number = 1, advancedOptions = undefined) => async (dispatch, getState) => {
  try {
    const assetHolderDAO = daoByType('TimeHolder')(getState())
    const { account } = getState().get(DUCK_SESSION)
    advancedOptions['account'] = account
    const tx: TxExecModel = await assetHolderDAO.withdraw(token, amount, feeMultiplier, advancedOptions)
  } catch (e) {
    // eslint-disable-next-line
    console.error('withdraw error', e)
  }
}
