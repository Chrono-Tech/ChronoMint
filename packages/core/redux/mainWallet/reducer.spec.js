/**
 * Copyright 2017â€“2018, LaborX PTY
 * Licensed under the AGPL Version 3 license.
 */

import Immutable from 'immutable'
import { accounts } from 'specsInit'
import MainWallet from '../../models/wallet/MainWalletModel'
import TransactionsCollection from '../../models/wallet/TransactionsCollection'
import TokenModel from '../../models/tokens/TokenModel'
import TxModel from '../../models/TxModel'
import * as a from './actions'
import reducer from './reducer'

const token1 = new TokenModel({ symbol: 'TK1' })

const tokens = new Immutable.Map({ [token1.symbol()]: token1 })

const tx1 = new TxModel({ txHash: 'hash1', from: 1, to: 2 })
const tx2 = new TxModel({ txHash: 'hash2', from: 3, to: 4 })

describe('settings wallet reducer', () => {
  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toEqual(new MainWallet())
  })

  // it('should handle WALLET_TOKENS', () => {
  //   const tokens = new Immutable.Map({
  //     [token1.symbol()]: token1,
  //     [token2.symbol()]: token2,
  //   })
  //   expect(reducer(new MainWallet({}), { type: a.WALLET_TOKENS, tokens })).toMatchSnapshot()
  // })

  it('should handle WALLET_BALANCE', () => {
    expect(reducer(new MainWallet({
      tokens,
    }), {
      type: a.WALLET_BALANCE, token: token1, isCredited: true, amount: 5,
    })).toMatchSnapshot()
  })

  it('should handle WALLET_ALLOWANCE', () => {
    expect(reducer(new MainWallet({
      tokens,
    }), {
      type: a.WALLET_ALLOWANCE, token: token1, spender: accounts[4], value: 4,
    })).toMatchSnapshot()
  })

  // it('should handle WALLET_TIME_DEPOSIT with non empty wallet', () => {
  //   expect(reducer(new MainWallet({
  //     timeDeposit: new BigNumber(5),
  //   }), {
  //     type: a.WALLET_TIME_DEPOSIT, isCredited: false, amount: 3,
  //   })).toMatchSnapshot()
  // })

  // it('should handle WALLET_TIME_DEPOSIT with empty wallet', () => {
  //   expect(reducer(new MainWallet(), { type: a.WALLET_TIME_ADDRESS, address: accounts[5] })).toMatchSnapshot()
  // })

  it('should handle WALLET_TRANSACTIONS_FETCH', () => {
    const initial = new MainWallet({
      transactions: new TransactionsCollection({
        list: new Immutable.Map({ a: 1 }),
        isFetching: false,
      }),
    })
    expect(reducer(initial, { type: a.WALLET_TRANSACTIONS_FETCH })).toMatchSnapshot()
  })

  it('should handle WALLET_TRANSACTION', () => {
    const initial = new MainWallet({
      transactions: new TransactionsCollection({
        list: new Immutable.Map({
          [tx1.id()]: tx1,
        }),
      }),
    })
    const updatedTx = new TxModel({
      txHash: 'hash1', from: 1, to: 2, blockNumber: 10,
    })

    expect(reducer(initial, { type: a.WALLET_TRANSACTION, tx: updatedTx })).toMatchSnapshot()
  })

  it('should handle WALLET_TRANSACTIONS', () => {
    const initial = new MainWallet({
      transactions: new TransactionsCollection({
        list: new Immutable.Map({
          tx1,
        }),
        endOfList: true,
        isFetching: true,
      }),
    })
    expect(reducer(initial, { type: a.WALLET_TRANSACTIONS, map: { tx2 } })).toMatchSnapshot()
  })

  it('should handle WALLET_IS_TIME_REQUIRED', () => {
    expect(reducer(new MainWallet({ isTIMERequired: true }), { type: a.WALLET_IS_TIME_REQUIRED, value: false })).toMatchSnapshot()
  })
})
