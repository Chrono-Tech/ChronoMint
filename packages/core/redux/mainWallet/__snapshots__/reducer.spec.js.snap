// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`settings wallet reducer should handle WALLET_BALANCE 1`] = `
{
  "address": null,
  "addresses": {
    "emptyModel": {
      "address": null,
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isInited": false,
      "isPending": false,
      "isSelected": false,
      "timestamp": null,
      "transactionHash": null
    },
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  },
  "allowances": {
    "emptyModel": {
      "amount": "0",
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isInited": false,
      "isPending": false,
      "isSelected": false,
      "spender": null,
      "timestamp": null,
      "token": null,
      "transactionHash": null
    },
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  },
  "balances": {
    "emptyModel": {
      "amount": "0",
      "id": null,
      "timestamp": null
    },
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {
      "null": {
        "amount": "5",
        "id": null,
        "timestamp": null
      }
    },
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  },
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isInited": false,
  "isMainWallet": true,
  "isMultisig": false,
  "isPending": false,
  "isSelected": false,
  "isTIMERequired": true,
  "timeAddress": null,
  "timestamp": null,
  "tokens": {
    "TK1": {
      "address": null,
      "balance": "0",
      "blockchain": null,
      "dao": null,
      "decimals": 0,
      "fee": {
        "fee": null,
        "feeAddress": null,
        "id": null,
        "isFailed": false,
        "isFetched": false,
        "isFetching": false,
        "isInited": false,
        "isPending": false,
        "isSelected": false,
        "timestamp": null,
        "transactionHash": null,
        "withFee": null
      },
      "feeRate": null,
      "icon": null,
      "id": null,
      "isERC20": false,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isInited": false,
      "isLocked": false,
      "isOptional": true,
      "isPending": false,
      "isReissuable": {
        "id": null,
        "isFailed": false,
        "isFetched": false,
        "isFetching": false,
        "isInited": false,
        "isPending": false,
        "isSelected": false,
        "timestamp": null,
        "transactionHash": null,
        "value": null
      },
      "isSelected": false,
      "managersList": {
        "emptyModel": {
          "address": null,
          "id": null,
          "isFailed": false,
          "isFetched": false,
          "isFetching": false,
          "isInited": false,
          "isPending": false,
          "isSelected": false,
          "timestamp": null,
          "transactionHash": null
        },
        "id": null,
        "isFailed": false,
        "isFetched": false,
        "isFetching": false,
        "isInited": false,
        "isPending": false,
        "isSelected": false,
        "leftToFetch": 0,
        "list": {},
        "selected": null,
        "timestamp": null,
        "transactionHash": null
      },
      "name": null,
      "platform": null,
      "symbol": "TK1",
      "timestamp": null,
      "totalSupply": "0",
      "transactionHash": null,
      "url": null
    }
  },
  "transactionHash": null,
  "transactions": {
    "emptyModel": {
      "args": null,
      "blockHash": null,
      "blockNumber": null,
      "by": null,
      "credited": null,
      "fee": "0",
      "from": null,
      "gas": null,
      "gasFee": "0",
      "gasPrice": null,
      "id": null,
      "input": null,
      "symbol": "",
      "time": null,
      "timestamp": null,
      "to": null,
      "token": null,
      "tokenAddress": null,
      "transactionIndex": null,
      "txHash": null,
      "type": "",
      "value": "0"
    },
    "endOfList": false,
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  }
}
`;

exports[`settings wallet reducer should handle WALLET_IS_TIME_REQUIRED 1`] = `
{
  "address": null,
  "addresses": {
    "emptyModel": {
      "address": null,
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isInited": false,
      "isPending": false,
      "isSelected": false,
      "timestamp": null,
      "transactionHash": null
    },
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  },
  "allowances": {
    "emptyModel": {
      "amount": "0",
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isInited": false,
      "isPending": false,
      "isSelected": false,
      "spender": null,
      "timestamp": null,
      "token": null,
      "transactionHash": null
    },
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  },
  "balances": {
    "emptyModel": {
      "amount": "0",
      "id": null,
      "timestamp": null
    },
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  },
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isInited": false,
  "isMainWallet": true,
  "isMultisig": false,
  "isPending": false,
  "isSelected": false,
  "isTIMERequired": false,
  "timeAddress": null,
  "timestamp": null,
  "tokens": {},
  "transactionHash": null,
  "transactions": {
    "emptyModel": {
      "args": null,
      "blockHash": null,
      "blockNumber": null,
      "by": null,
      "credited": null,
      "fee": "0",
      "from": null,
      "gas": null,
      "gasFee": "0",
      "gasPrice": null,
      "id": null,
      "input": null,
      "symbol": "",
      "time": null,
      "timestamp": null,
      "to": null,
      "token": null,
      "tokenAddress": null,
      "transactionIndex": null,
      "txHash": null,
      "type": "",
      "value": "0"
    },
    "endOfList": false,
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  }
}
`;

exports[`settings wallet reducer should handle WALLET_TRANSACTION 1`] = `
{
  "address": null,
  "addresses": {
    "emptyModel": {
      "address": null,
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isInited": false,
      "isPending": false,
      "isSelected": false,
      "timestamp": null,
      "transactionHash": null
    },
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  },
  "allowances": {
    "emptyModel": {
      "amount": "0",
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isInited": false,
      "isPending": false,
      "isSelected": false,
      "spender": null,
      "timestamp": null,
      "token": null,
      "transactionHash": null
    },
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  },
  "balances": {
    "emptyModel": {
      "amount": "0",
      "id": null,
      "timestamp": null
    },
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  },
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isInited": false,
  "isMainWallet": true,
  "isMultisig": false,
  "isPending": false,
  "isSelected": false,
  "isTIMERequired": true,
  "timeAddress": null,
  "timestamp": null,
  "tokens": {},
  "transactionHash": null,
  "transactions": {
    "emptyModel": {
      "args": null,
      "blockHash": null,
      "blockNumber": null,
      "by": null,
      "credited": null,
      "fee": "0",
      "from": null,
      "gas": null,
      "gasFee": "0",
      "gasPrice": null,
      "id": null,
      "input": null,
      "symbol": "",
      "time": null,
      "timestamp": null,
      "to": null,
      "token": null,
      "tokenAddress": null,
      "transactionIndex": null,
      "txHash": null,
      "type": "",
      "value": "0"
    },
    "endOfList": false,
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {
      "hash1 - 1 - 2": {
        "args": null,
        "blockHash": null,
        "blockNumber": 10,
        "by": null,
        "credited": null,
        "fee": "0",
        "from": 1,
        "gas": null,
        "gasFee": "0",
        "gasPrice": null,
        "id": null,
        "input": null,
        "symbol": "",
        "time": null,
        "timestamp": null,
        "to": 2,
        "token": null,
        "tokenAddress": null,
        "transactionIndex": null,
        "txHash": "hash1",
        "type": "",
        "value": "0"
      }
    },
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  }
}
`;

exports[`settings wallet reducer should handle WALLET_TRANSACTIONS 1`] = `
{
  "address": null,
  "addresses": {
    "emptyModel": {
      "address": null,
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isInited": false,
      "isPending": false,
      "isSelected": false,
      "timestamp": null,
      "transactionHash": null
    },
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  },
  "allowances": {
    "emptyModel": {
      "amount": "0",
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isInited": false,
      "isPending": false,
      "isSelected": false,
      "spender": null,
      "timestamp": null,
      "token": null,
      "transactionHash": null
    },
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  },
  "balances": {
    "emptyModel": {
      "amount": "0",
      "id": null,
      "timestamp": null
    },
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  },
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isInited": false,
  "isMainWallet": true,
  "isMultisig": false,
  "isPending": false,
  "isSelected": false,
  "isTIMERequired": true,
  "timeAddress": null,
  "timestamp": null,
  "tokens": {},
  "transactionHash": null,
  "transactions": true
}
`;

exports[`settings wallet reducer should handle WALLET_TRANSACTIONS_FETCH 1`] = `
{
  "address": null,
  "addresses": {
    "emptyModel": {
      "address": null,
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isInited": false,
      "isPending": false,
      "isSelected": false,
      "timestamp": null,
      "transactionHash": null
    },
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  },
  "allowances": {
    "emptyModel": {
      "amount": "0",
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isInited": false,
      "isPending": false,
      "isSelected": false,
      "spender": null,
      "timestamp": null,
      "token": null,
      "transactionHash": null
    },
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  },
  "balances": {
    "emptyModel": {
      "amount": "0",
      "id": null,
      "timestamp": null
    },
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  },
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isInited": false,
  "isMainWallet": true,
  "isMultisig": false,
  "isPending": false,
  "isSelected": false,
  "isTIMERequired": true,
  "timeAddress": null,
  "timestamp": null,
  "tokens": {},
  "transactionHash": null,
  "transactions": {
    "emptyModel": {
      "args": null,
      "blockHash": null,
      "blockNumber": null,
      "by": null,
      "credited": null,
      "fee": "0",
      "from": null,
      "gas": null,
      "gasFee": "0",
      "gasPrice": null,
      "id": null,
      "input": null,
      "symbol": "",
      "time": null,
      "timestamp": null,
      "to": null,
      "token": null,
      "tokenAddress": null,
      "transactionIndex": null,
      "txHash": null,
      "type": "",
      "value": "0"
    },
    "endOfList": false,
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": true,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {
      "a": 1
    },
    "selected": null,
    "timestamp": null,
    "transactionHash": null
  }
}
`;
